src/lib/communication/DY.Lib.Communication.Core.Lemmas.fst

```diff
@@ -0,0 +1,573 @@
+  in
+  let sig_input_bytes = serialize signature_input sig_input in
+  serialize_wf_lemma signature_input (is_publishable tr) sig_input;
+  let signature = sign sk_sender nonce sig_input_bytes in
+  let msg_signed = SigMessage {msg=sig_input_bytes; signature} in
+  assert(bytes_invariant tr signature);
+  serialize_wf_lemma com_message_t (is_publishable tr) msg_signed;
+  ()
+
+val send_authenticated_proof:
+  {|invs:protocol_invariants|} ->
+  #a:Type -> {| parseable_serializeable bytes a |} ->
+  tr:trace ->
+  higher_layer_preds:comm_higher_layer_event_preds a ->
+  comm_keys_ids:communication_keys_sess_ids ->
+  sender:principal -> receiver:principal -> payload:a ->
+  Lemma
+  (requires
+    trace_invariant tr /\
+    has_private_keys_invariant /\
+    has_communication_layer_crypto_predicates /\
+    has_communication_layer_event_predicates higher_layer_preds /\
+    higher_layer_preds.send_auth tr sender payload /\
+    is_well_formed a (is_publishable tr) payload
+  )
+  (ensures (
+    let (_, tr_out) = send_authenticated #a comm_keys_ids sender receiver payload tr in
+    trace_invariant tr_out
+  ))
+  [SMTPat (trace_invariant #invs tr);
+   SMTPat (send_authenticated comm_keys_ids sender receiver payload tr);
+   SMTPat (apply_core_comm_layer_lemmas higher_layer_preds)]
+let send_authenticated_proof #invs #a tr higher_layer_preds comm_keys_ids sender receiver payload =
+    reveal_opaque (`%send_authenticated) (send_authenticated #a);
+    match send_authenticated comm_keys_ids sender receiver payload tr with
+    | (None, tr_out) -> ()
+    | (Some _, tr_out) -> (
+      let (Some sk_sender, tr') = get_private_key sender comm_keys_ids.private_keys (LongTermSigKey comm_layer_sign_tag) tr in
+      let (nonce,  tr') = mk_rand SigNonce (long_term_key_label sender) 32 tr' in
+      higher_layer_preds.send_auth_later tr tr' sender payload;
+      let payload_bytes = serialize a payload in
+      let ((), tr') = trigger_event sender (CommAuthSendMsg sender payload_bytes) tr' in
+      assert(event_triggered tr' sender (CommAuthSendMsg sender payload_bytes));
+      sign_message_proof #invs.crypto_invs #a tr' sender receiver payload None sk_sender nonce;
+      let msg_signed = sign_message #a sender receiver payload None sk_sender nonce in
+      let (msg_id, tr') = send_msg msg_signed tr' in
+      assert(tr_out == tr');
+      assert(trace_invariant tr_out);
+      ()
+    )
+
+
+#push-options "--z3rlimit 50"
+val verify_message_proof:
+  {|cinvs:crypto_invariants|} ->
+  #a:Type -> {| ps:parseable_serializeable bytes a |} ->
+  tr:trace ->
+  sender:principal -> receiver:principal ->
+  msg_bytes:bytes -> sk_receiver_opt:option bytes ->
+  vk_sender:bytes ->
+  Lemma
+  (requires
+    has_communication_layer_crypto_predicates /\
+    is_publishable tr msg_bytes /\
+    is_public_key_for tr vk_sender (LongTermSigKey comm_layer_sign_tag) sender
+  )
+  (ensures (
+    reveal_opaque (`%verify_message) (verify_message #a);
```

Comment:
Why should we reveal the definition of `verify_message` here?
(I guess we don't need to anymore this its `opaque_to_smt` attribute was commented out though)

Reply:
The `ensures` of this lemma implicitly relies on the guarantee that if `verify_message` is `Some`, then `msg_bytes` is parseable, has a specific form, and so on. When `verify_message` is opaque and not revealed, the `ensures` clause therefore fails to check. For now, this should probably also be removed, and get added on to the set of things to fix when making `verify_message` opaque.

Reply:
Now removed (though I may add an issue to try to clean this up further in the future).